tensorflow/compiler/xla/service/gpu/miopen_conv_algorithm_picker.cc

```diff
@@ -224,45 +303,57 @@ MiopenConvAlgorithmPicker::PickBestAlgorithm(
           &stream, ShapeUtil::ByteSizeOf(instr->shape().tuple_shapes(0))));
   initialize_buffer(result_buffer);
 
-  se::dnn::ProfileResult best_result;
-  int64 best_result_bytes_used = 0;
-  TF_ASSIGN_OR_RETURN(auto backend_config,
-                      instr->backend_config<CudnnConvBackendConfig>());
+  TF_ASSIGN_OR_RETURN(std::vector<AlgorithmDesc> algorithms,
+                      GetAlgorithms(instr, absl::MakeSpan(operand_buffers),
+                                    result_buffer, stream_exec_, &stream));
+
+  std::vector<MiopenConvAlgorithmPicker::AutotuneResult> profile_results;
+
+  for (const AlgorithmDesc& alg : algorithms) {
+    XLA_SCOPED_LOGGING_TIMER_LEVEL(
+        absl::StrCat("CudnnConvAlgorithmPicker::PickBestAlgorithm algo ",
+                     AlgorithmToString(alg)),
+        2);
+
+    ScratchAllocator scratch_allocator(device_ordinal, allocator);
+    se::dnn::ProfileResult profile_result;
+    VLOG(3) << "Trying algorithm " << AlgorithmToString(alg) << " for "
+            << instr->ToString();
 
-  ScratchAllocator scratch_allocator(device_ordinal, allocator);
-  se::dnn::ProfileResult profile_result;
-  VLOG(3) << "Auto-tuning for " << instr->ToString();
-  RunConvOptions options;
-  options.profile_result = &profile_result;
-  options.first_call_from_algorithm_picker = true;
-  bool launch_ok =
+    // Use assignment instead of brace-list to make GCC 4.9 happy.
+    RunConvOptions options;
+    options.profile_result = &profile_result;
+    options.algo_override = alg;
+    Status launch_status =
         RunCudnnConv(instr, absl::MakeSpan(operand_buffers), result_buffer,
```

Comment:
Something I just realized is that in each profiling round, we are allocating memory for that algorithm. I *think* we might be able to do better by. (Might be applicable in find mode as well):

1. Query the worst case scenario of memory consumption (like in find-mode)
2. Allocate once and only once the worst case scenario of memory
3. Reuse the whole chunk of memory by all profiling rounds

Let me know if you think this make sense or not